src/client.ts

```diff
@@ -569,6 +600,102 @@ in an environmental variable named FAUNA_SECRET or pass it to the Client\
   }
 }
 
+export type StreamEventHandler = (event: StreamEvent) => void;
+
+export class StreamClient {
+  #callbacks: Record<StreamEventType, StreamEventHandler[]> = {
+    start: [],
+    add: [],
+    remove: [],
+    update: [],
+    error: [],
+  };
+  #query: StreamToken;
+  #clientConfiguration: Record<string, any>;
+  #httpStreamClient: HTTPStreamClient;
+  #streamAdapter: StreamAdapter | null = null;
+
+  // TODO: make clientConfiguration and httpStreamClient optional
+  constructor(
+    query: StreamToken,
+    clientConfiguration: StreamClientConfiguration,
+    httpStreamClient: HTTPStreamClient
+  ) {
+    this.#query = query;
+    this.#clientConfiguration = clientConfiguration;
+    this.#httpStreamClient = httpStreamClient;
+  }
+
+  on(type: StreamEventType, callback: StreamEventHandler) {
+    this.#callbacks[type].push(callback);
+    return this;
+  }
+
+  start() {
+    const headers = {
+      Authorization: `Bearer ${this.#clientConfiguration.secret}`,
+    };
+
+    const streamAdapter = this.#httpStreamClient.stream({
+      data: this.#query.token,
+      headers,
+      method: "POST",
+    });
+
+    this.#streamAdapter = streamAdapter;
+
+    const handleEvents = async () => {
+      for await (const event of streamAdapter.read) {
+        // stream events are always tagged
+        const deserializedEvent: StreamEvent = TaggedTypeFormat.decode(event, {
+          long_type: this.#clientConfiguration.long_type,
+        });
+        this.#callbacks[deserializedEvent.type].forEach((callback) =>
+          callback(deserializedEvent)
+        );
+      }
+    };
+
+    handleEvents().catch((error) => {
+      throw new NetworkError("Error reading from stream", { cause: error });
+    });
```

Comment:
how would this trigger in the user application code? where would I add my error handling for this scenario?

Reply:
You have to put `stream.start()` or `for await (event of stream.iter())` in a try/catch block.

Reply:
Actually I don't think that this Network error would be caught by a try catch block around `stream.start()` and the callback method. I think we actually would need to pipe this into the error callback. And then I don't know what to do if the user doesn't register one. That's a problem that also exists with the v4 driver.